### DISCUSSION TITLE
Understanding hooks in Pinia-ORM and issues with saving/creating Instances

### DISCUSSION THREAD

**Discussion by sergerdn** [CodeDredd/pinia-orm#1106] [General]
I completely do not understand how hooks like `saving` and `creating` work. I want to create some hooks to populate my model fields with custom logic in the following scenarios:
- Before creating the instance
- Before saving the instance
- Before updating the instance.

I have a hook named `creating` that populates models with default fields using my logic, and a hook named `saving` with some logic too. Please note that the code I provided is only a simple example. In reality, I have more complex logic that I need to implement.

```javascript 
import { Model } from "pinia-orm";
import { Attr, Str, Uid } from "pinia-orm/dist/decorators";

class User extends Model {
    static entity = "user_profiles";

    @Uid() declare id: string;
    @Str("") declare name: string;

    @Attr("") declare createdAt: Date;
    @Attr("") declare updatedAt: Date;

    static creating(model: Model) {
        console.log("creating");
        if (!(model.createdAt instanceof Date) || isNaN(model.createdAt.getTime())) {
            model.createdAt = new Date();
        }

        if (!model.name) {
            model.name = "set some name";
        }
    }

    static updating(model: Model) {
        console.log("updating");
    }

    static saving(model: Model) {
        console.log("saving");
        // Set default values for updatedAt field
        model.updatedAt = new Date();
    }
}

export { User };
```

What I did:
1. When I use `const user = useRepo(User).new()`, all of my hooks fire but the models are saved, but I don't need them to be saved. **This seems to fulfill all of my needs, but without saving the instance.**
2. When I use `const user = useRepo(User).make()`, none of my hooks seem to fire, even though I expect the `creating` hook to be fired.

I am trying to find a way to create a new instance of a model in Pinia-ORM without saving it, while still firing the `creating` hook, similar to using `useRepo(User).new()`. Is there a way to accomplish this using Pinia-ORM?

P.S.
I have read all the documentation, issues, Google searches, etc. multiple times over the past two days, but I didn't find any instructions.

```json
{ 
    "dependencies": {
        "@mdi/font": "^7.2.96",
        "pinia": "^2.0.32",
        "pinia-orm": "^1.5.1",
        "vue": "^3.2.47",
        "vue-router": "^4.1.6",
        "vuetify": "^3.1.10"
    },
    "devDependencies": {
        "@rushstack/eslint-patch": "^1.2.0",
        "@types/jsdom": "^21.1.0",
        "@types/node": "^18.14.2",
        "@vitejs/plugin-vue": "^4.0.0",
        "@vitejs/plugin-vue-jsx": "^3.0.0",
        "@vue/eslint-config-prettier": "^7.1.0",
        "@vue/eslint-config-typescript": "^11.0.2",
        "@vue/test-utils": "^2.3.0",
        "@vue/tsconfig": "^0.1.3",
        "concurrently": "^7.6.0",
        "cypress": "^12.7.0",
        "electron": "^23.1.4",
        "electron-builder": "^23.6.0",
        "eslint": "^8.34.0",
        "eslint-plugin-cypress": "^2.12.1",
        "eslint-plugin-vue": "^9.9.0",
        "jsdom": "^21.1.0",
        "npm-run-all": "^4.1.5",
        "prettier": "^2.8.4",
        "start-server-and-test": "^2.0.0",
        "typescript": "~4.9.5",
        "vite": "^4.1.4",
        "vitest": "^0.29.1",
        "vue-tsc": "^1.2.0",
        "wait-on": "^7.0.1"
    }
}
```

**Comment by CodeDredd** [comment]
Help is on the way #1107. Righ now it's not possible to have what you need. Even if you use `return false` in the hooks then you just get null. But you surly want the model instance.

You can thank me if you like with a cup of ☕.

**Reply by sergerdn** [reply]
Wow, thank you @CodeDredd for your amazing support!

Just so you know, I'm pretty new to JavaScript and I was thinking that I just didn't find anything that I needed, instead of thinking that those things are not available yet.

I would like to buy you a coffee, but I am currently facing some issues. Nevertheless, I will try to solve them. GitHub has informed me that there are some errors with my payments. I am using a paid GitHub account, and I have never seen such an error before.

**Reply by CodeDredd** [reply]
Just realeased for you the fix https://github.com/CodeDredd/pinia-orm/releases/tag/pinia-orm%401.6.1 . It's fine. Maybe somehow you get it to work with sending me a coffee 😄

**Reply by sergerdn** [reply]
Thank you very much, I will test it as soon as possible. Unfortunately, GitHub has banned me from buying a coffee for you due to economic restrictions. Therefore, my payment failed.😪

But in any case, I would still like to help and address the issue. I have created a ticket to the GitHub support team. Once it is resolved, if possible, I will be more than happy to buy you a coffee.

**Reply by sergerdn** [reply]
I have tested it and it works perfectly. However, I have a question. If I create an object with `const user = useRepo(User).new(false)`, the object is returned with an `id` value. So, how can I tell if the object has been saved or just created for future reference?

Maybe we should not generate an `id` if the object has just been created and not yet saved?

I asked because someone has a fairly extensive application, and in certain instances, they would like to determine if an object has been saved or simply created in an easy way.

What do you think?

**Reply by CodeDredd** [reply]
Hmmm....i get your point. But then you need to use just `@Attr` for id. Then you will have the outcome you want. `@UId` will always create an id because maybe you use it for an not id field.

I am still thinking if this easy solveable or needed at all since you can define ids as not uid. But it kind of feels right to make it that way you say it.

**Reply by sergerdn** [reply]
I believe that using `@Attr` for the primary key is not a problem in most cases. I have only used `@UId` in testing scenarios, and I will change it to `@Attr` because it will help me with relations. Currently, the relations are not working as expected and documented with `@UId`, but that is another story.

**Comment by sergerdn** [comment]
It seems like we have a problem - the `saving` hook is being fired when calling `userProfileRepo.new(false)`, which I did not expect. Since the object is not being saved, I assumed that the `saving` hook would only be triggered when the object is actually saved, before the save operation takes place.


`User.ts`:
```javascript
import { Model } from "pinia-orm"
import { Attr, Cast, Str, Uid } from "pinia-orm/dist/decorators"
import { DateCast } from "pinia-orm/dist/casts"

class User extends Model {
    static entity = "user_profiles"

    @Uid() declare id: string
    @Str("") declare name: string

    @Cast(() => DateCast) @Attr(null) declare createdAt: Date
    @Cast(() => DateCast) @Attr(null) declare updatedAt: Date

    static creating(model: Model) {
        console.log("creating")
        if (model.createdAt == null) {
            model.createdAt = new Date()
        }

        if (!model.name) {
            model.name = "set some name"
        }
    }

    static updating(model: Model) {
        console.log("updating")
        model.updatedAt = new Date()
    }

    static saving(model: Model) {
        console.log("saving")
        model.updatedAt = new Date()
    }
}

export { User }
```

I encountered an error that I did not expect:
```
AssertionError: expected 2023-03-23T11:42:06.965Z to be null
Expected :undefined
Actual   :2023-03-23T11:42:06.965Z
```

`User.spec.ts`:
```javascript
import { afterEach, beforeEach, describe, expect, it } from "vitest"
import { createPinia, type Pinia } from "pinia"
import { createORM, Model, Repository, useRepo } from "pinia-orm"

import { User } from "./User"

describe("User", () => {
    // Define some variables that will be used in the tests
    let pinia: Pinia
    let userProfileRepo: Repository<User>
    beforeEach(() => {
        pinia = createPinia().use(createORM())
        userProfileRepo = useRepo(User, pinia)
    })

    afterEach(async () => {
        Model.clearRegistries()
    })

    it("userProfileRepo.new(true)", () => {
        expect(userProfileRepo.all().length).toBe(0) // double check
        const user = userProfileRepo.new(true)!

        expect(userProfileRepo.all().length).toBe(1) //  should save
        expect(user.createdAt).toBeInstanceOf(Date)
        expect(user.updatedAt).toBeInstanceOf(Date)
        expect(user.name).toBe("set some name")
    })

   it("userProfileRepo.new(false), broken test", () => {
        expect(userProfileRepo.all().length).toBe(0) // double check
        const user = userProfileRepo.new(false)!
        expect(userProfileRepo.all().length).toBe(0) // should not save

        expect(user.createdAt).toBeInstanceOf(Date)
        expect(user.updatedAt).toBeNull()  // test broken is here
        expect(user.name).toBe("set some name")
    })
})
```

`package.json`:

```
{
    "dependencies": {
        "@mdi/font": "^7.2.96",
        "pinia": "^2.0.33",
        "pinia-orm": "^1.6.1",
        "vue": "^3.2.47",
        "vue-router": "^4.1.6",
        "vuetify": "^3.1.11"
    },
    "devDependencies": {
        "@rushstack/eslint-patch": "^1.2.0",
        "@types/jsdom": "^21.1.1",
        "@types/node": "^18.15.5",
        "@vitejs/plugin-vue": "^4.1.0",
        "@vitejs/plugin-vue-jsx": "^3.0.1",
        "@vue/eslint-config-prettier": "^7.1.0",
        "@vue/eslint-config-typescript": "^11.0.2",
        "@vue/test-utils": "^2.3.2",
        "@vue/tsconfig": "^0.1.3",
        "concurrently": "^7.6.0",
        "cypress": "^12.8.1",
        "electron": "^23.2.0",
        "electron-builder": "^23.6.0",
        "eslint": "^8.36.0",
        "eslint-plugin-cypress": "^2.12.1",
        "eslint-plugin-vue": "^9.9.0",
        "jsdom": "^21.1.1",
        "npm-check-updates": "^16.8.0",
        "npm-run-all": "^4.1.5",
        "prettier": "^2.8.6",
        "start-server-and-test": "^2.0.0",
        "typescript": "~5.0.2",
        "vite": "^4.2.1",
        "vitest": "^0.29.7",
        "vue-tsc": "^1.2.0",
        "wait-on": "^7.0.1"
    }
}
```

**Reply by CodeDredd** [reply]
Before Data is saved:

creating
updating
saving (which is triggered if a record is updating or creating)
And the afterr saved hooks:

created
updated
saved (triggered if a recrod is updated or saved)

Maybe i should this to the docs

**Reply by sergerdn** [reply]
I'm curious, why is `saving` triggered upon object creation? What is the rationale behind it?
```javascript
 const user = userProfileRepo.new(true)! // object creation only, we do not save anything
```

**Reply by CodeDredd** [reply]
`saving` is just a shortcut if need the same logic for `updating` or `creating`. 
I think the wording is just cofusing you. `saving` is more the synonym for `creatingOrUpdating`

**Reply by sergerdn** [reply]
> `saving` is just a shortcut if need the same logic for `updating` or `creating`.
> I think the wording is just cofusing you. `saving` is more the synonym for `creatingOrUpdating`

Wow. Now I see you. Is it documented or not?

**Reply by CodeDredd** [reply]
> Wow. Now I see you. Is it documented or not?

No it's not proper documented. I will add it. I come from the laravel world so this seemed clear for me ^^.

**Reply by sergerdn** [reply]
> `saving` is just a shortcut if need the same logic for `updating` or `creating`.
> I think the wording is just cofusing you. `saving` is more the synonym for `creatingOrUpdating`

Okay, but we have a problem here. Where is the hook named `hookBeforeSavingData` that specifically fires before the model is saved? I really need such a hook for my app because I need to do some things before the models are saved to the Pinia storage.

Take a look:
https://github.com/CodeDredd/pinia-orm/discussions/1106#discussioncomment-5420989

Maybe the hook named `saved` will be useful for me, but I'm not sure at the moment.

I believe I have a way to hook Pinia's storage before using the ORM to write data, although I'm not yet sure how to execute it. Still, I'm optimistic that it can be done.

**Reply by sergerdn** [reply]
I will think about it and write a proposal for you outlining the name and method for implementing the hook. In any case, I will do my best to ensure its success.

I will need some time to read through the documentation on how the other library was created and how they named their hooks.

**Reply by CodeDredd** [reply]
> Okay, but we have a problem here. Where is the hook named hookBeforeSavingData that specifically fires before the model is saved? I really need such a hook for my app because I need to do some things before the models are saved to the Pinia storage.

This hook is `creating` where you also return `false` to prevent saving to the store. All hooks `creating`, `updating`, `saving` & `deleting` are called before the store is updated.

**Reply by sergerdn** [reply]
> This hook is `creating` where you also return `false` to prevent saving to the store. All hooks `creating`, `updating`, `saving` & `deleting` are called before the store is updated.

I also do not understand how returning `false` works in `creation` hooks. It seems like it discards the instance creation.

`User.ts`
```javascript 
import { Model } from "pinia-orm"
import { Attr, Str, Uid } from "pinia-orm/dist/decorators"

class User extends Model {
    static entity = "user_profiles"

    @Uid() declare id: string
    @Str("") declare name: string

    @Attr(null) declare createdAt: number
    @Attr(null) declare updatedAt: number

    static creating(model: Model) {
        console.log("creating")
        if (model.createdAt == null) {
            model.createdAt = new Date().getTime()
        }

        if (!model.name) {
            model.name = "set some name"
        }

        return false
    }

    static updating(model: Model) {
        console.log("updating")
        model.updatedAt = new Date().getTime()
    }

    static saving(model: Model) {
        console.log("saving")
        model.updatedAt = new Date().getTime()
    }
}

export { User }
```

I returned `false` in the `creating` hook, and I did not expect to see the message `saving`. Also, `userProfileRepo.new(false)` does not return an instance of the model for me.

```
creating
saving

AssertionError: expected null to be truthy
```


```javascript
import { afterEach, beforeEach, describe, expect, it } from "vitest"
import { createPinia, type Pinia } from "pinia"
import { createORM, Model, Repository, useRepo } from "pinia-orm"

import { User } from "./User"

describe("User", () => {
    // Define some variables that will be used in the tests
    let pinia: Pinia
    let userProfileRepo: Repository<User>
    beforeEach(() => {
        pinia = createPinia().use(createORM())
        userProfileRepo = useRepo(User, pinia)
    })

    afterEach(async () => {
        Model.clearRegistries()
    })

    it("userProfileRepo.new(false)", () => {
        expect(userProfileRepo.all().length).toBe(0) // double check
        // hook creating fired because new instance
        const user = userProfileRepo.new(false)!
        expect(user).toBeTruthy() // ERROR IS HEREEE

        expect(userProfileRepo.all().length).toBe(0)

        // hook creating MUST NOT fire because NOT new instance
        userProfileRepo.save(user)
        expect(userProfileRepo.all().length).toBe(1)

        // Assert that the createdAt and updatedAt fields are set correctly.
        expect(new Date(user?.createdAt!)).toBeInstanceOf(Date)
        expect(new Date(user?.updatedAt!)).toBeInstanceOf(Date)
        expect(user.name).toBe("set some name")
    })
})
```

**Reply by sergerdn** [reply]
I believe that naming and implementing the hooks's logic in a different way might be better. 
Here's a proposal:

We have hooks named `before`:
- `beforeCreated`, called before instance is created
- `beforeSaved`, called before instance is saved
- `beforeUpdated`, called before instance is updated

We also have hooks named `after`:
- `afterCreated`, called after instance is created
- `afterSaved`, called after instance is saved
- `afterUpdated`, called after instance is updated

Additionally, we could have optional hooks named `do` to replace the current logic:
- `doCreated`, to replace `created` logic
- `doSaved`, to replace `saved` logic
- `doUpdated`, to replace `updated` logic

Some use cases:

1. If I create a new instance with `const user = useRepo(User).new(false)`, the `beforeCreated` hook should be fired because it is a new model instance. The `beforeSaved` hook should not be fired because the model has not been saved yet. Additionally, the `beforeUpdated` hook should not be fired because it is a completely new model.

2. If I create a new instance with `const user = useRepo(User).new(true)`, hooks should be fired: `beforeCreated`, `beforeSaved`, because the model is being saved and created. The `beforeUpdated` hook should not be fired because it is a completely new model.

3. If I update my instance, the `beforeUpdated` hook should fire because I am updating the model. It does not matter how the instance was created because the update operation implies that the instance already exists. The `beforeSaved` hook should not fire because updating the model does not necessarily mean it will be saved. 

P.S.
Having optional hooks named `do` to replace the current logic could be interesting, but it is important to note that it may introduce confusion. For example, if we replace the `doSaved` hook with "nothing", the model will never be saved, which could be very funny 😀.

**Comment by sergerdn** [comment]
I also encountered a strange error with another test, and I'm not sure if it's related to the new feature or not.

`User.ts`(same):
```javascript
import { Model } from "pinia-orm"
import { Attr, Cast, Str, Uid } from "pinia-orm/dist/decorators"
import { DateCast } from "pinia-orm/dist/casts"

class User extends Model {
    static entity = "user_profiles"

    @Uid() declare id: string
    @Str("") declare name: string

    @Cast(() => DateCast) @Attr(null) declare createdAt: Date
    @Cast(() => DateCast) @Attr(null) declare updatedAt: Date

    static creating(model: Model) {
        console.log("creating")
        if (model.createdAt == null) {
            model.createdAt = new Date()
        }

        if (!model.name) {
            model.name = "set some name"
        }
    }

    static updating(model: Model) {
        console.log("updating")
        model.updatedAt = new Date()
    }

    static saving(model: Model) {
        console.log("saving")
        model.updatedAt = new Date()
    }
}

export { User }
```

```
TypeError: (intermediate value)(intermediate value)(intermediate value).toISOString is not a function
    at DateCast.set (node_modules/pinia-orm/dist/casts.mjs:107:78)
.....
    at Array.map (<anonymous>)
```
It happens on a code:
```javascript
class DateCast extends CastAttribute {
  /**
   * Create a new String attribute instance.
   */
  constructor(attributes) {
    super(attributes);
  }
  get(value) {
    return value === null ? null : new Date(value);
  }
  /**
   * Make the value for the attribute.
   */
  set(value) {
    if (value === null)
      return null;
    return (typeof value === "string" ? new Date(Date.parse(value)) : value).toISOString(); // <<<<<== HERE
  }
}
```

`UserAnotherTest.spec.ts`:

```javascript
import { afterEach, beforeEach, describe, expect, it } from "vitest"
import { createPinia, type Pinia } from "pinia"
import { createORM, Model, Repository, useRepo } from "pinia-orm"

import { User } from "./User"

describe("User", () => {
    // Define some variables that will be used in the tests
    let pinia: Pinia
    let userProfileRepo: Repository<User>

    beforeEach(() => {
        pinia = createPinia().use(createORM())
        userProfileRepo = useRepo(User, pinia)
    })

    afterEach(async () => {
        Model.clearRegistries()
    })

    it("userProfileRepo.new(true): error with toISOString is not a function", () => {
        expect(userProfileRepo.all().length).toBe(0) // double check
        const user = userProfileRepo.new(true)!
        expect(userProfileRepo.all().length).toBe(1)

        expect(user.createdAt).toBeInstanceOf(Date)
        expect(user.updatedAt).toBeInstanceOf(Date)
        expect(user.name).toBe("set some name")
      
        userProfileRepo.where("id", user.id).update({ name: "Modified name" }) ///<< ERROR IS HERE
    })

    it("userProfileRepo.new(false): error with toISOString is not a function", () => {
        expect(userProfileRepo.all().length).toBe(0) // double check
        const user = userProfileRepo.new(false)!
        expect(userProfileRepo.all().length).toBe(0)

        expect(user.createdAt).toBeInstanceOf(Date)
        expect(user.updatedAt).toBeInstanceOf(Date)
        expect(user.name).toBe("set some name")

        userProfileRepo.save(user)
        expect(userProfileRepo.all().length).toBe(1)

        userProfileRepo.where("id", user.id).update({ name: "Modified name" }) ///<< ERROR IS HERE

    })
})
```

**Reply by CodeDredd** [reply]
Well i think that's a design flaw which you found. The problem is that the casts are triggered before the hooks. Maybe in future i can somehow change that. For now you should simply use this if you have casts in the model:

````ts
import { Model } from "pinia-orm"
import { Attr, Cast, Str, Uid } from "pinia-orm/dist/decorators"
import { DateCast } from "pinia-orm/dist/casts"

class User extends Model {
    static entity = "user_profiles"

    @Uid() declare id: string
    @Str("") declare name: string

    @Cast(() => DateCast) @Attr(null) declare createdAt: Date
    @Cast(() => DateCast) @Attr(null) declare updatedAt: Date

    static creating(model: User) {
        console.log("creating")
        if (model.createdAt == null) {
            // @ts-expect-error Cast is not triggered in the hook
            model.createdAt = (new Date()).toISOString()
        }

        if (!model.name) {
            model.name = "set some name"
        }
    }

    static updating(model: User) {
        console.log("updating")
        // @ts-expect-error Cast is not triggered in the hook
        model.updatedAt = (new Date()).toISOString()
    }

    static saving(model: User) {
        console.log("saving")
        // @ts-expect-error Cast is not triggered in the hook
        model.updatedAt = (new Date()).toISOString()
    }
}

export { User }
````

i am going to make a PR for this

**Reply by sergerdn** [reply]
I figured it out by setting `createdAt` as a timestamp number. I believe this will help me avoid some issues in the future. Perhaps using a timestamp instead of `Date` would be a better choice for me. 

```javascript
class User extends Model {
    @Attr(null) declare createdAt: number 
    @Attr(null) declare updatedAt: number
}

static creating(model: BrowserProfile) {
        console.log("creating: ", model)
        model.createdAt = new Date().getTime() // timestamp
```

**Reply by CodeDredd** [reply]
Well i anyway making the change since it makes totally sense to trigger casts before some is saved into the store. Also for hooks. Now you can use both.
https://github.com/CodeDredd/pinia-orm/pull/1113 😄 

....so just released an update. On the run i encountered some hydration bugs. So this topic helped finding them 👍

**Reply by sergerdn** [reply]
Have you added tests for models with fields like this?
```javascript
class User extends Model {
    @Cast(() => DateCast) @Attr(null) declare createdAt: Date
    @Cast(() => DateCast) @Attr(null) declare updatedAt: Date
}
```

I debugged some of your code and wrote to you about it, but perhaps there was a misunderstanding. Maybe I didn't express myself very clearly.

```
class DateCast extends CastAttribute {
// ......
  set(value) {
    if (value === null)
      return null;

    //  HERE HERREEEEE, Because we received something that we were not expecting. << LOOOK :-)
    return (typeof value === "string" ? new Date(Date.parse(value)) : value).toISOString(); 
  }
}
``

**Reply by sergerdn** [reply]
> . So this topic helped finding them 👍

It's okay for me to find bugs anywhere in my life.

**Reply by CodeDredd** [reply]
> I debugged some of your code and wrote to you about it, but perhaps there was a misunderstanding. Maybe I didn't express myself very clearly.

Sry i never thought that somehow and empty objet is created. I don't know yet why. But i adapted DateCast so it defently will work. You can try out v1.6.3

**Reply by CodeDredd** [reply]
FOUND it 🎉 ....released again and fix. Now everything is correct. Try v1.6.4

**Reply by CodeDredd** [reply]
Btw... I don't know if you know it and if it is waht you need. But there is meta field if you need `updatedAt` and `createdAt`. https://pinia-orm.codedredd.de/guide/model/getting-started#meta-info

**Reply by sergerdn** [reply]
> FOUND it 🎉 ....released again and fix. Now everything is correct. Try v1.6.4

Yeah, I took a look at https://github.com/CodeDredd/pinia-orm/commit/519681c258885110478191e9bfd3a9dfb016d670. I'm sorry, I should have been more clear in my previous messages. 

However, I have a problem. I'm not very experienced with JavaScript, which can sometimes make it difficult for other people to understand how I describe certain things.

**Reply by CodeDredd** [reply]
It's all ok 😉 .... sometimes i get self confused by my own words 😆

**Reply by sergerdn** [reply]
> Btw... I don't know if you know it and if it is waht you need. But there is meta field if you need `updatedAt` and `createdAt`. https://pinia-orm.codedredd.de/guide/model/getting-started#meta-info

Yes, I have seen it. However, there are certain situations where I prefer to rely on my own logic so that I can better understand and meet my needs. This is especially true for my app, which will both receive and send data to an Electron app.

``` md
         Vue.js App                      DataProfileModel (read storage) 
             │                                   │
             │   Write data to                   │
             ▼   Electron API                    ▼
     dataProfileModelPlugin.ts <───IPC───>      Electron
             │                                   │
             │    Read and write data to         │
             ▼    Electron State                 ▼
        Pinia ORM State                     Electron State
             │                                   │
             │    Read data from                 │
             ▼    DataProfileModel               ▼
      DataProfileModel                      DataProfileModel (write to storage)
```

Vue.js app communicates with Pinia ORM to manage data. Pinia ORM communicates with DataProfileModel to read/write data. `dataProfilePlugin.ts` bridges Vue.js app and Electron using IPC for CRUD operations on DataProfileModel. Electron updates Electron state, which syncs with Pinia ORM state. Updates to Pinia ORM state are reflected in Vue.js app's UI.
